x-pack/plugins/lens/public/xy_visualization_plugin/xy_expression.tsx

```diff
@@ -0,0 +1,245 @@
+    return {
+      type: 'legendConfig',
+      ...args,
+    };
+  },
+};
+
+/**
+ * Properties common to all axes.
+ */
+interface AxisConfig {
+  title: string;
+  showGridlines: boolean;
+  position: Position;
+}
+
+/**
+ * AxisConfig expression definition.
+ */
+const axisConfig = {
+  title: { types: ['string'] },
+  showGridlines: { types: ['boolean'] },
+  position: { types: ['string'] },
+};
+
+/**
+ * Y-axis configuration.
+ */
+export interface YConfig extends AxisConfig {
+  accessors: string[];
+}
+
+/**
+ * YConfig as an expression type.
+ */
+export const yConfig = {
+  name: 'yConfig',
+  aliases: [],
+  type: 'yConfig',
+  help: `Configure a chart's y axis`,
+  context: {
+    types: ['null'],
+  },
+  args: {
+    ...axisConfig,
+    accessors: {
+      types: ['string'],
+      multi: true,
+    },
+  },
+  fn: function fn(_context: any, args: YConfig) {
+    return {
+      type: 'yConfig',
+      ...args,
+    };
+  },
+};
+
+/**
+ * X-axis configuration.
+ */
+export interface XConfig extends AxisConfig {
+  accessor: string;
+}
+
+/**
+ * XConfig as an expression type.
+ */
+export const xConfig = {
+  name: 'xConfig',
+  aliases: [],
+  type: 'xConfig',
+  help: `Configure a chart's x axis`,
+  context: {
+    types: ['null'],
+  },
+  args: {
+    ...axisConfig,
+    accessor: { types: ['string'] },
+  },
+  fn: function fn(_context: any, args: XConfig) {
+    return {
+      type: 'xConfig',
+      ...args,
+    };
+  },
+};
+
+/**
+ * The arguments to the XY chart expression function.
+ */
+export interface XYArgs {
+  seriesType: 'bar' | 'line' | 'area';
+  title: string;
+  legend: LegendConfig;
+  y: YConfig;
+  x: XConfig;
+  splitSeriesAccessors: string[];
+  stackAccessors: string[];
+}
+
+/**
+ * The XY chart expression function.
+ */
+export const xyChart = {
+  name: 'xy_chart',
+  type: 'render',
+  args: {
+    seriesType: { types: ['string'] },
+    title: { types: ['string'] },
+    legend: { types: ['legendConfig'] },
+    y: { types: ['yConfig'] },
+    x: { types: ['xConfig'] },
+    splitSeriesAccessors: {
+      types: ['string'],
+      multi: true,
+    },
+    stackAccessors: {
+      types: ['string'],
+      multi: true,
+    },
+  },
+  context: { types: ['kibana_datatable'] },
+  fn(data: Datatable, args: XYArgs) {
+    return {
+      type: 'render',
+      as: 'xy_chart_renderer',
+      value: {
+        data,
+        args,
+      },
+    };
+  },
+};
+
+/**
+ * The XY chart expression renderer.
+ */
+export const xyChartRenderer = {
+  name: 'xy_chart_renderer',
+  displayName: 'XY Chart',
+  reuseDomNode: true,
+  render: async (domNode: HTMLDivElement, config: any, _handlers: any) => {
+    ReactDOM.render(<XYChart {...config} />, domNode);
+  },
+};
+
+/**
+ * Render an XY chart as a React component.
+ *
+ * @param props - The data and args which are used to render the chart.
+ */
+function XYChart({ data, args }: { data: Datatable; args: XYArgs }) {
```

Comment:
Can we also test the actual chart? I vaguely remember something from the charts channel that there is a problem with that.